### PR TITLE
Attachment Carousel Virtual Caching

### PR DESCRIPTION
**Pull request by kidroca**
### Details

Add a virtual list to the carousel component that is used to display attachments in the message view.
This allows us to render (and preload) a large number of attachments without having to render them all at once.
Now navigating between attachments is instant because we preload the next attachment in the background.

### Fixed Issues

$ https://github.com/Expensify/App/issues/15922
PROPOSAL: https://github.com/Expensify/App/issues/15922#issuecomment-1483358712

#### Preparation
1. Open a report with a lot of attachments (I used the #announce channel from ECX)
2. Open an attachment

#### Things to check
- [ ] Navigate between attachments using the left and right arrows works
- [ ] If we wait for a second before pressing "Next Arrow" - the attachment is preloaded when you navigate to it
- [ ] Going back to a previous attachment is instant (they are loaded from cache almost instantly)
- [ ] On mobile devices (including mobile web) swiping works, and we can navigate between attachments
- [ ] Open an attachment from the "middle" works (not the first or last attachment)
    - scroll to the past and select an attachment
    - the attachment view / carousel should open with the correct attachment
    - navigating between attachments should move to neighboring attachments
- [ ] On mobile (including mobile web), taping on the carousel should toggle between showing and hiding the navigation arrows 
- [ ] Downloading attachment still works and downloads the correct attachment

### Offline tests

Offline behavior is same as before:
- images would appear to be loading forever (visible spinner)
- Images that were seen got cached and should appear even when offline 

### QA Steps

The following steps are for both STAGING and PRODUCTION environments and all platforms 

#### Preparation
1. Open a report with a lot of attachments (I used the #announce channel from ECX)
2. Open an attachment

#### Things to check
- [ ] Navigate between attachments using the left and right arrows works
- [ ] If we wait for a second before pressing "Next Arrow" - the attachment is preloaded when you navigate to it
- [ ] Going back to a previous attachment is instant (they are loaded from cache almost instantly)
- [ ] On mobile devices (including mobile web) swiping works, and we can navigate between attachments
- [ ] Open an attachment from the "middle" works (not the first or last attachment)
    - scroll to the past and select an attachment
    - the attachment view / carousel should open with the correct attachment
    - navigating between attachments should move to neighboring attachments
- [ ] On mobile (including mobile web), taping on the carousel should toggle between showing and hiding the navigation arrows 
- [ ] Downloading attachment still works and downloads the correct attachment

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/12156624/230182861-f2aa5adb-69c5-4807-ac26-ceb29ee9f3ba.mp4

https://user-images.githubusercontent.com/12156624/230439932-f51dce6a-8785-4e10-89ae-a7021da9edf4.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/12156624/230182129-1740cdeb-6ab3-4a6d-9cd3-c7f0e215ed9d.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/12156624/230440007-edb97570-3161-4376-b531-e1604e7910e7.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/12156624/230439922-2fb8f2d8-7dc2-4308-9222-f3499cfc713e.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/12156624/230439966-29b513ec-b777-426e-a6f9-270e1dcc7b0e.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/12156624/230181928-23918400-1571-4490-bbcc-f939e8f9e35e.mp4

</details>

### CODE REVIEW

File: src/components/AttachmentCarousel/index.js [hunk 1]
```diff
@@ -124,50 +154,95 @@ class AttachmentCarousel extends React.Component {
...
+        }
+
+        return (
+            <Pressable
+                // eslint-disable-next-line react/jsx-props-no-spreading
+                {...props}
+                onPress={() => this.setState(current => ({shouldShowArrow: !current.shouldShowArrow}))}
+                style={style}
+            />
+        );
```
Comment (kidroca) [inline src/components/AttachmentCarousel/index.js:220]: This is something I picked up while reviewing the `onPress` of `CarouselActions` (now removed)

File: src/components/AttachmentCarousel/index.js [hunk 2]
```diff
@@ -197,20 +272,37 @@ class AttachmentCarousel extends React.Component {
...
                 )}
-                <CarouselActions
-                    styles={[styles.attachmentModalArrowsContainer]}
-                    canSwipeLeft={!this.state.isBackDisabled}
-                    canSwipeRight={!this.state.isForwardDisabled}
-                    onPress={() => this.canUseTouchScreen && this.toggleArrowsVisibility(!this.state.shouldShowArrow)}
-                    onCycleThroughAttachments={this.cycleThroughAttachments}
-                >
-                    <AttachmentView
-                        onPress={() => this.toggleArrowsVisibility(!this.state.shouldShowArrow)}
```
Comment (kidroca) [inline src/components/AttachmentCarousel/index.js:208]: Follow up refactor - the `onPress` prop seem to have been added to the `AttchmentView` and then all the child view that it renders (Image, Pdf, ...) for the arrow functionality 

There are no more usages of AttachmentView with `onPress`, so we can remove it (as well as the other views like ImageView)

**Review by kidroca — COMMENTED** [review]
Added some notes on the changes

**Comment by kidroca** [comment]
> * It's not possible to scroll the pages of pdf file

Which platform - I don't have a problem on iOS, Chrome, Desktop, mobile web. I'll test Android tomorrow

> * If we swipe left or right with two fingers on mac trackpad the attachment will shake

This is because of the snap to alignment feature, I'll look into it

> * If we zoom an image and go to the next image then go back, the previous image remains in its zoomed-in position

This is from successful virtualization - the image doesn't reload, it's properties aren't changed it's just moved into view. 
If we want to fix this we'll have to come up with some special logic that rests only the zoom level when items scroll out of view

For such a small bug I wouldn't risk adding complex logic to try and fix it

**Comment by kidroca** [comment]
> * It's not possible to scroll the pages of pdf file
> * If we swipe left or right with two fingers on mac trackpad the attachment will shake

I've pushed a fix for these 2 items 

---

Regarding the issue with resetting zoom 

> * If we zoom an image and go to the next image then go back, the previous image remains in its zoomed-in position

I found out how to rest zoom for images, but not for PDF documents 

I would prefer to address this in a separate PR as it would require a bit more changes and testing 
Like an `isVisible` prop added to the `AttachmentView` which would become `false` when item is no longer in view
and based on that the component can run logic to (partially) reset itself 

Let's first confirm whether we want to fix this at all 
(BTW: If you switch to 2-3 images away and go back to the zoomed image it is not zoomed anymore)

---

✅ Ready for Review

**Comment by johnmlee101** [comment]
> Let's first confirm whether we want to fix this at all 

The zoom seems fine, I don't think it's much of a "bug" to have it retain the previous state of when you were viewing it

**Comment by kidroca** [comment]
> When I increase the screen size I can see the image list and sometimes the image changes. @kidroca, can we fix this?

I don't think there's anything easy to do about it - the app is laggy when it's resized like that
For a blink the window becomes wider, before that information is relayed to App state, so components like FlatList can't know they're showing more than they have to

**Review by thesahindia — APPROVED** [review]
Works well!

cc: @johnmlee101

**Review by johnmlee101 — APPROVED** [review]
This worked very well! Thanks for this

**Pull request merged**